src/NHibernate/Dialect/Oracle12cDialect.cs

```diff
@@ -0,0 +1,54 @@
+using NHibernate.SqlCommand;
+using NHibernate.SqlCommand.Parser;
+
+namespace NHibernate.Dialect
+{
+	/// <summary> 
+	/// A dialect specifically for use with Oracle 10g.
+	/// </summary>
+	/// <remarks>
+	/// The main difference between this dialect and <see cref="Oracle12cDialect"/>
+	/// is the use of "ANSI join syntax" here...
+	/// </remarks>
+	public class Oracle12cDialect : Oracle10gDialect
+	{
+		/// <summary>
+		/// Oracle 12c supports a query statement that provides <c>LIMIT</c>
+		/// functionality with an offset.
+		/// </summary>
+		/// <value><c>false</c></value>
+		public override bool UseMaxForLimit
+		{
+			get { return false; }
+		}
+
+		public override SqlString GetLimitString(SqlString querySqlString, SqlString offset, SqlString limit)
+		{
+			var tokenEnum = new SqlTokenizer(querySqlString).GetEnumerator();
+			if (!tokenEnum.TryParseUntilFirstMsSqlSelectColumn()) return null;
+
+			var result = new SqlStringBuilder(querySqlString);
+			if (!tokenEnum.TryParseUntil("order"))
+			{
+				result.Add(" ORDER BY CURRENT_TIMESTAMP");
```

Comment:
This has been added to Sql Server dialect, because it does not allow to use new fetching syntax without ordering results.